pkg/store/proxy.go

```diff
@@ -384,14 +394,34 @@ func startStreamSeriesSet(
 			}
 		}()
 		for {
-			r, err := s.stream.Recv()
+			frameTimeoutCtx, cancel := startFrameCtx(s.responseTimeout)
+			if cancel != nil {
+				defer cancel()
+			}
+			rCh := make(chan *recvResponse, 1)
+			var rr *recvResponse
+			go func() {
```

Comment:
So we care about the first frame right? or timeout for all frames?

Reply:
In any case I think we need to start this go routine before `for` loop and have here `for` loop as well.

This will make sure that we only have one 2 go routines running: one waiting for recv or context cancel, second for reading.

Current implementation will constantly allocate new channel and go routine. For 1000 frames x 100 concurrent queries this might matter.


I wish we have benchmarks for querier ))): Like we do for Store now: https://github.com/thanos-io/thanos/blob/88f6be883b66036c589c5697034e3d99c0848b3b/pkg/store/bucket_test.go#L1071

Actually added issue: https://github.com/thanos-io/thanos/issues/2105